Change: When building NewGRFs, use grfid (from grfcodec) to calculate NewGRF MD5 hash.

PeterN opened the pull request:
## Motivation / Problem

#11491 adds 32bpp sprites to the extra data NewGRF, which requires it to be built as container v2, but our build scripts do not correctly calculate the MD5 hash for container v2 NewGRFs.

`Baseset.cmake` uses CMake's built in MD5 function to calculate the hash of `orig_extra.grf`. However hash calculation only works for v1 container grfs.

The tool `grfid` from grfcodec can be used to calculate both v1 and v2 hashes, however grfcodec isn't currently a build dependency.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

<s>Vendor grfid (with some changes due to our updated Md5 implementation and use of fmt, and use it in the build process.

Switch to NewGRF container v2. A side effect of this is that the files are a about 10% larger.</s>

Update the build system to use grfid from grfcodec to calculate the hash of the custom NewGRFs. This requires a bit more effect than before as MD5 was a nice cmake built-in. The hash is written to a file which is stored in the source tree to enable OpenTTD to be compiled if grfcodec is not present.


<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

<s>I removed the copyright message on grfid -- it's my name but I never put that there, so...</s>

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

glx22 commented on src/grfid/grfid.cpp:
```diff
@@ -0,0 +1,289 @@
+#else
+		this->file_buffer = new uint8_t[this->file_length];
+		if (this->file_buffer == nullptr) return false;
+		size_t file_read = read(this->fd, this->file_buffer, this->file_length);
```
MSVC warnings are valid (the function expects a 32bit unsigned int) and `file_length` is 64bit.
But `lseek` return is a 32bit signed value, so an explicit cast should be fine.

PeterN replied:
Who would ever have a 2GB file... but okay, unlikely for a NewGRF :D

glx22 commented on media/baseset/CMakeLists.txt:
```diff
@@ -30,6 +30,7 @@ if(GRFCODEC_FOUND)
 
 set(BASESET_EXTRAGRF_FILE ${CMAKE_CURRENT_SOURCE_DIR}/orig_extra.grf)
+set(BASESET_EXTRAGRF_HASH ${CMAKE_CURRENT_SOURCE_DIR}/orig_extra.grf.md5)
 
```
```suggestion

if (HOST_BINARY_DIR)
    include(${HOST_BINARY_DIR}/grfid.cmake)
endif()
```

glx22 commented on media/baseset/CMakeLists.txt:
```diff
@@ -38,17 +39,25 @@ foreach(BASESET_SOURCE_FILE IN LISTS BASESET_SOURCE_FILES)
+        COMMAND grfid -m ${BASESET_EXTRAGRF_FILE} > ${BASESET_EXTRAGRF_HASH}
+        DEPENDS grfid ${BASESET_EXTRAGRF_FILE}
+    )
+
```
And I'd extract this outside of the `foreach`

PeterN committed:
Change: When building NewGRFs, use grfid (from grfcodec) to calculate NewGRF MD5 hash.

This allows for v2 container NewGRFs to be created. Hashes are needed for building the baseset metadata, so the .hash files are stored the source tree so they can be built if grfcodec and grfid are not present.

The hash of openttd.grf is not actually needed, but it is simpler to leave it in than handle only orig_extra.grf.

glx22 commented on cmake/scripts/CreateGRF.cmake:
```diff
@@ -55,4 +58,18 @@ if(RESULT)
+
+# Trim hash from grfid
+string(STRIP ${GRFID_HASH} GRFID_HASH)
+file(WRITE ${CMAKE_SOURCE_DIR}/${GRF_SOURCE_FOLDER_NAME}.grf.hash ${GRFID_HASH})
```
`${CMAKE_SOURCE_DIR}` is current working dir for script ran with `cmake -P`
But 
```suggestion
file(WRITE ${GRF_BINARY_FILE}.hash ${GRFID_HASH})
```
should work without needing the manual copy.

PeterN replied:
Oops, that was me trying things that didn't do anything, and not reverting it :)

glx22 commented on cmake/CreateGrfCommand.cmake:
```diff
@@ -42,9 +43,14 @@ function(create_grf_command)
+    add_custom_target(${GRF_SOURCE_FOLDER_NAME}.grf.hash
+            DEPENDS
+            ${GRF_BINARY_FILE}
+    )
```
This target will do the same as `${GRF_SOURCE_FOLDER_NAME}.grf` target, I don't think it's needed to add it.
